### PR TITLE
santad/sysx: Prevent unlinking databases

### PR DESCRIPTION
**Pull request by russellhancox**

### CODE REVIEW

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm [hunk 1]
```diff
@@ -240,6 +240,32 @@ - (void)messageHandler:(es_message_t *)m API_AVAILABLE(macos(10.15)) {
...
     }
+    case ES_EVENT_TYPE_AUTH_UNLINK: {
+      es_string_token_t pathToken = m->event.unlink.target->path;
+      NSString *path = [[NSString alloc] initWithBytes:pathToken.data
+                                                length:pathToken.length
+                                              encoding:NSUTF8StringEncoding];
+      if ([path isEqualToString:@"/private/var/db/santa/rules.db"] ||
+          [path isEqualToString:@"/private/var/db/santa/events.db"]) {
+        LOGW(@"Preventing attempt to delete Santa databases!");
+        es_respond_auth_result(self.client, m, ES_AUTH_RESULT_DENY, false);
```
Comment (tburgin) [inline Source/santad/EventProviders/SNTEndpointSecurityManager.mm:251]: I think we can cache this, we are never going to allow unlinking or renaming for these paths.

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm [hunk 2]
```diff
@@ -240,6 +240,32 @@ - (void)messageHandler:(es_message_t *)m API_AVAILABLE(macos(10.15)) {
...
+      }
+      es_respond_auth_result(self.client, m, ES_AUTH_RESULT_ALLOW, false);
```
Comment (tburgin) [inline Source/santad/EventProviders/SNTEndpointSecurityManager.mm:253]: Same for the allow.

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm (hunk 1 above)
Comment (russellhancox) [reply Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: Good point

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm (hunk 2 above)
Comment (russellhancox) [reply Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: Done - same for rename

**Comment by russellhancox** [comment]
PTAL: also added a check to allow santad to unlink/rename if it needs to.

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm [hunk 3]
```diff
@@ -240,6 +240,35 @@ - (void)messageHandler:(es_message_t *)m API_AVAILABLE(macos(10.15)) {
...
       break;
     }
+    case ES_EVENT_TYPE_AUTH_UNLINK: {
+      es_string_token_t pathToken = m->event.unlink.target->path;
+      NSString *path = [[NSString alloc] initWithBytes:pathToken.data
+                                                length:pathToken.length
+                                              encoding:NSUTF8StringEncoding];
+      if (([path isEqualToString:@"/private/var/db/santa/rules.db"] ||
+           [path isEqualToString:@"/private/var/db/santa/events.db"]) &&
+           audit_token_to_pid(m->process->audit_token) != getpid()) {
```
Comment (tburgin) [inline Source/santad/EventProviders/SNTEndpointSecurityManager.mm:250]: nit: self.selfPID

Comment (russellhancox) [reply Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: Done

**Pull request merged**